modules/internal/CPtr.chpl

```diff
@@ -223,8 +223,26 @@ module CPtr {
     :arg arr: the array for which we should retrieve a pointer
     :returns: a pointer to the array data
   */
-  inline proc c_ptrTo(arr: []) where isRectangularArr(arr) {
-    if !arr._value.oneDData then halt("error: c_ptrTo(multi_ddata array");
+  inline proc c_ptrTo(arr: []) where isRectangularArr(arr) && !arr._value.isDefaultRectangular() {
+    if !chpl__getActualArray(arr).oneDData then halt("error: c_ptrTo(multi_ddata array");
+    return c_pointer_return(arr[arr.domain.low]);
+  }
+
+  pragma "no doc"
+  inline proc c_ptrTo(arr: []) where arr._value.isDefaultRectangular() {
+    const val = arr._value;
+    if chpl__isArrayView(val) {
```

Comment:
Both, I think (though I may be misunderstanding what checks you're referring to).  I was imagining something like:

````chapel
proc c_ptrTo(arr: []) {
  return _array.cPtrTo();
}

proc _array.cPtrTo() {
  return _value.dsiCPtrTo();
}

// and then a per domain-map dsiCPtrTo() routine that would either generate an error or return the pointer
````

The idea being to get rid of a lot of where clauses and cases in this c_ptrTo() routine and have each domain map answer the question itself.